File: lambdas/pstn-transfer-ts/src/httpclient.ts
/**
 * HTTP client with retry logic matching the Go implementation
 */

import type { OAuth2TokenFetcher } from "./auth.js";
import type { Logger } from "./logger.js";
import { getDurationFromEnv, getIntFromEnv } from "./utils.js";

const HTTP_MAX_RETRIES = getIntFromEnv("HTTP_MAX_RETRIES", 3);
const HTTP_RETRY_BASE_DELAY = getDurationFromEnv("HTTP_RETRY_BASE_DELAY", 100);
const HTTP_CLIENT_TIMEOUT = getDurationFromEnv("HTTP_CLIENT_TIMEOUT", 10000);

export interface HTTPClient {
	fetch(request: Request): Promise<Response>;
}

export interface AuthConfig {
	apiKey?: string; // Deprecated
	authDomain?: string; // Auth domain without path (e.g., "https://auth.us-west-2-prod.cresta.ai")
	oauthClientID?: string;
	oauthClientSecret?: string;
	tokenFetcher?: OAuth2TokenFetcher;
}

/**
 * Checks if an error is an abort/cancellation error
 */
function isAbortError(err: unknown): boolean {
	if (!(err instanceof Error)) {
		return false;
	}
	return (
		err.name === "AbortError" ||
		err.name === "AbortedError" ||
		/aborted|cancelled|canceled/i.test(err.message)
	);
}

/**
 * Creates a timeout signal that combines the parent signal with a timeout
 * Returns both the signal and a cleanup function
 */
function withTimeoutSignal(
	parent: AbortSignal | null | undefined,
	timeoutMs: number,
): { signal: AbortSignal; cleanup: () => void } {
	const controller = new AbortController();
	let timeoutId: ReturnType<typeof setTimeout> | null = null;
	let abortListener: (() => void) | null = null;

	// Set up timeout
	timeoutId = setTimeout(() => {
		controller.abort(new Error(`request timeout after ${timeoutMs}ms`));
	}, timeoutMs);

	// Listen to parent signal if provided
	if (parent) {
		if (parent.aborted) {
			controller.abort(parent.reason);
		} else {
			abortListener = () => {
				controller.abort(parent.reason);
			};
			parent.addEventListener("abort", abortListener, { once: true });
		}
	}

	const cleanup = () => {
		if (timeoutId !== null) {
			clearTimeout(timeoutId);
			timeoutId = null;
		}
		if (parent && abortListener) {
			parent.removeEventListener("abort", abortListener);
			abortListener = null;
		}
	};

	return { signal: controller.signal, cleanup };
}

export class RetryHTTPClient implements HTTPClient {
	private logger?: Logger;
	private authConfig?: AuthConfig;
	private maxRetries: number;
	private baseDelay: number;

	constructor(options?: {
		logger?: Logger;
		authConfig?: AuthConfig;
		maxRetries?: number;
		baseDelay?: number;
	}) {
		this.logger = options?.logger;
		this.authConfig = options?.authConfig;
		this.maxRetries = options?.maxRetries ?? HTTP_MAX_RETRIES;
		this.baseDelay = options?.baseDelay ?? HTTP_RETRY_BASE_DELAY;
	}

	async fetch(request: Request): Promise<Response> {
		// Check if AbortSignal is already aborted
		if (request.signal?.aborted) {
			throw new Error(`context cancelled: ${request.signal.reason}`);
		}

		// Read request body into memory so we can recreate it for retries
		let bodyBytes: ArrayBuffer | null = null;
		if (request.body) {
			// Clone the request to read body without consuming the original
			const clonedRequest = request.clone();
			bodyBytes = await clonedRequest.arrayBuffer();
		}

		let lastErr: Error | null = null;
		let lastResp: Response | null = null;

		for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
			if (attempt > 0) {
				// Check if AbortSignal is aborted before retrying
				if (request.signal?.aborted) {
					throw new Error(`context cancelled: ${request.signal.reason}`);
				}

				const delay = exponentialBackoff(attempt - 1, this.baseDelay);
				if (this.logger) {
					this.logger.debugf(
						"Retrying request to %s (attempt %d/%d) after %dms",
						request.url,
						attempt + 1,
						this.maxRetries + 1,
						delay,
					);
				}

				await new Promise((resolve) => setTimeout(resolve, delay));
			}

			// Recreate request preserving all original fields, only overriding body
			// Use new Request(request, {...}) to preserve: credentials, mode, cache, redirect, etc.
			const retryRequest = new Request(request, {
				body: bodyBytes ?? null,
				// Don't set signal here - we'll use the combined timeout signal
			});

			// Add authentication header if configured
			// Use retryRequest (not original request) for consistency
			if (this.authConfig) {
				try {
					const authHeader = await this.getAuthHeader(retryRequest);
					if (authHeader && !retryRequest.headers.get("Authorization")) {
						retryRequest.headers.set("Authorization", authHeader);
					}
				} catch (authErr) {
					// Auth failures should not be retried - fail fast
					const authError = authErr instanceof Error ? authErr : new Error(String(authErr));
					throw new Error(`error getting auth header: ${authError.message}`);
				}
			}

			// Create combined timeout signal with proper cleanup
			const { signal: timeoutSignal, cleanup } = withTimeoutSignal(
				request.signal ?? null,
				HTTP_CLIENT_TIMEOUT,
			);

			try {
				const response = await fetch(retryRequest, {
					signal: timeoutSignal,
				});

				// Check if status code is retryable
				if (!isRetryableError(null, response.status)) {
					// Non-retryable, return immediately
					return response;
				}

				// Retryable status code - consume body and prepare for retry
				// Consume body to release resources
				try {
					await response.arrayBuffer();
				} catch (bodyErr) {
					// If we can't consume the body, log but continue
					if (this.logger) {
						this.logger.debugf(
							"Error consuming response body: %v",
							bodyErr instanceof Error ? bodyErr.message : String(bodyErr),
						);
					}
				}

				// Store response and error for final error message
				lastResp = response;
				lastErr = new Error(`request returned retryable status: ${response.status}`);
			} catch (err) {
				const error = err instanceof Error ? err : new Error(String(err));

				// Don't retry on abort/cancellation errors - fail immediately
				if (isAbortError(error)) {
					throw error;
				}

				// Check if this error should be retried
				if (!isRetryableError(error, 0)) {
					// Non-retryable error (e.g., programming errors, auth failures)
					throw new Error(`error making HTTP request: ${error.message}`);
				}

				// Retryable error - store and continue
				lastErr = error;
			} finally {
				// Always cleanup timeout and event listeners
				cleanup();
			}
		}

		// All retries exhausted
		if (lastErr) {
			throw new Error(`request failed after ${this.maxRetries + 1} attempts: ${lastErr.message}`);
		}
		if (lastResp) {
			throw new Error(
				`request failed after ${this.maxRetries + 1} attempts with status: ${lastResp.status}`,
			);
		}
		throw new Error(`request failed after ${this.maxRetries + 1} attempts`);
	}

	private async getAuthHeader(request: Request): Promise<string> {
		if (!this.authConfig) {
			throw new Error("authConfig is required");
		}

		// OAuth 2 authentication takes precedence
		if (this.authConfig.oauthClientID && this.authConfig.oauthClientSecret) {
			if (!this.authConfig.tokenFetcher) {
				throw new Error("tokenFetcher is required for OAuth authentication");
			}
			if (!this.authConfig.authDomain) {
				throw new Error("authDomain is required for OAuth authentication");
			}

			const token = await this.authConfig.tokenFetcher.getToken(
				request.signal ?? new AbortController().signal,
				this.authConfig.authDomain,
				this.authConfig.oauthClientID,
				this.authConfig.oauthClientSecret,
			);
			return `Bearer ${token}`;
		}

		// Fall back to API key authentication (deprecated)
		if (this.authConfig.apiKey) {
			return `ApiKey ${this.authConfig.apiKey}`;
		}

		throw new Error("no authentication configured");
	}
}

/**
 * isRetryableError determines if an error or status code should trigger a retry
 * Matches the Go implementation: retries network errors (err != nil), 5xx status codes,
 * 429 (Too Many Requests), and 408 (Request Timeout)
 */
export function isRetryableError(err: Error | null, statusCode: number): boolean {
	if (err) {
		// Network errors are retryable (but abort errors are filtered out before this)
		return true;
	}
	// Retry on 5xx server errors, 429 (Too Many Requests), and 408 (Request Timeout)
	return (statusCode >= 500 && statusCode < 600) || statusCode === 429 || statusCode === 408;
}

/**
 * exponentialBackoff calculates the delay for the given attempt with jitter
 */
export function exponentialBackoff(attempt: number, baseDelayMs: number): number {
	const delay = 2 ** attempt * baseDelayMs;
	// Add jitter: random value between 0 and 25% of delay
	const jitter = Math.random() * 0.25 * delay;
	return delay + jitter;
}

File: lambdas/pstn-transfer-py/src/httpclient.py
"""
HTTP client with retry logic matching the Go and TypeScript implementations
"""

from __future__ import annotations

import random
import time
import urllib.error
import urllib.request
from abc import ABC, abstractmethod
from typing import TYPE_CHECKING

from .types import AuthConfig
from .utils import get_duration_from_env, get_int_from_env

if TYPE_CHECKING:
    from .logger import Logger

# Configuration from environment
HTTP_MAX_RETRIES = get_int_from_env("HTTP_MAX_RETRIES", 3)
HTTP_RETRY_BASE_DELAY = get_duration_from_env("HTTP_RETRY_BASE_DELAY", 100)
HTTP_CLIENT_TIMEOUT = get_duration_from_env("HTTP_CLIENT_TIMEOUT", 10000)


class HTTPClient(ABC):
    """Abstract HTTP client interface"""

    @abstractmethod
    def fetch(
        self,
        method: str,
        url: str,
        headers: dict[str, str],
        body: bytes | None = None,
    ) -> bytes:
        """Make HTTP request and return response body"""
        pass


def is_retryable_error(err: Exception | None, status_code: int) -> bool:
    """
    Determines if an error or status code should trigger a retry.
    Matches the Go implementation: retries network errors (err != None), 5xx status codes,
    429 (Too Many Requests), and 408 (Request Timeout)
    """
    if err:
        # Network errors are retryable
        return True
    # Retry on 5xx server errors, 429 (Too Many Requests), and 408 (Request Timeout)
    return (500 <= status_code < 600) or status_code == 429 or status_code == 408


def exponential_backoff(attempt: int, base_delay_ms: int) -> int:
    """Calculates the delay for the given attempt with jitter"""
    delay = (2**attempt) * base_delay_ms
    # Add jitter: random value between 0 and 25% of delay
    jitter = random.random() * 0.25 * delay
    return int(delay + jitter)


class RetryHTTPClient(HTTPClient):
    """HTTP client with retry logic"""

    def __init__(
        self,
        logger: Logger | None = None,
        auth_config: AuthConfig | None = None,
        max_retries: int | None = None,
        base_delay: int | None = None,
    ) -> None:
        self._logger = logger
        self._auth_config = auth_config
        self._max_retries = max_retries if max_retries is not None else HTTP_MAX_RETRIES
        self._base_delay = base_delay if base_delay is not None else HTTP_RETRY_BASE_DELAY

    def fetch(
        self,
        method: str,
        url: str,
        headers: dict[str, str],
        body: bytes | None = None,
    ) -> bytes:
        """Make HTTP request with retry logic"""
        last_err: Exception | None = None
        last_status: int | None = None

        for attempt in range(self._max_retries + 1):
            if attempt > 0:
                delay = exponential_backoff(attempt - 1, self._base_delay)
                if self._logger:
                    self._logger.debugf(
                        "Retrying request to %s (attempt %d/%d) after %dms",
                        url,
                        attempt + 1,
                        self._max_retries + 1,
                        delay,
                    )
                time.sleep(delay / 1000.0)  # Convert ms to seconds

            # Add authentication header if configured
            request_headers = headers.copy()
            if self._auth_config:
                try:
                    auth_header = self._get_auth_header()
                    if auth_header and "Authorization" not in request_headers:
                        request_headers["Authorization"] = auth_header
                except Exception as auth_err:
                    # Auth failures should not be retried - fail fast
                    raise ValueError(f"error getting auth header: {auth_err}") from auth_err

            try:
                request = urllib.request.Request(
                    url,
                    data=body,
                    headers=request_headers,
                    method=method,
                )

                timeout_seconds = HTTP_CLIENT_TIMEOUT / 1000.0

                with urllib.request.urlopen(request, timeout=timeout_seconds) as response:
                    status_code = response.status
                    response_body: bytes = response.read()

                    # Check if status code is retryable
                    if not is_retryable_error(None, status_code):
                        # Non-retryable, return immediately
                        if status_code != 200:
                            raise ValueError(
                                f"request returned non-200 status: {status_code}, body: {response_body.decode('utf-8')}"
                            )
                        return response_body

                    # Retryable status code
                    last_status = status_code
                    last_err = ValueError(f"request returned retryable status: {status_code}")

            except urllib.error.HTTPError as e:
                status_code = e.code
                error_body = e.read().decode("utf-8") if e.fp else ""

                # Check if status code is retryable
                if not is_retryable_error(None, status_code):
                    # Non-retryable error
                    raise ValueError(
                        f"request returned non-200 status: {status_code}, body: {error_body}"
                    ) from e

                # Retryable status code
                last_status = status_code
                last_err = ValueError(f"request returned retryable status: {status_code}")

            except urllib.error.URLError as e:
                # Network error - retryable
                last_err = e

            except TimeoutError as e:
                # Timeout - retryable
                last_err = e

            except Exception as e:
                # Check if this error should be retried
                if not is_retryable_error(e, 0):
                    raise ValueError(f"error making HTTP request: {e}") from e
                last_err = e

        # All retries exhausted
        if last_err:
            raise ValueError(f"request failed after {self._max_retries + 1} attempts: {last_err}")
        if last_status:
            raise ValueError(
                f"request failed after {self._max_retries + 1} attempts with status: {last_status}"
            )
        raise ValueError(f"request failed after {self._max_retries + 1} attempts")

    def _get_auth_header(self) -> str:
        """Get authentication header value"""
        if not self._auth_config:
            raise ValueError("authConfig is required")

        # OAuth 2 authentication takes precedence
        if self._auth_config.oauth_client_id and self._auth_config.oauth_client_secret:
            if not self._auth_config.token_fetcher:
                raise ValueError("tokenFetcher is required for OAuth authentication")
            if not self._auth_config.auth_domain:
                raise ValueError("authDomain is required for OAuth authentication")

            token = self._auth_config.token_fetcher.get_token(
                self._auth_config.auth_domain,
                self._auth_config.oauth_client_id,
                self._auth_config.oauth_client_secret,
            )
            return f"Bearer {token}"

        # Fall back to API key authentication (deprecated)
        if self._auth_config.api_key:
            return f"ApiKey {self._auth_config.api_key}"

        raise ValueError("no authentication configured")
